### DISCUSSION TITLE
Unrealistic results in 3D simulation

### DISCUSSION THREAD

**Discussion by RWTHLHK** [idaholab/moose#23676] [Q&A Modules: Solid mechanics] [answered]
I got a weird result in a 3D simulation. The sample is a common notched round bar. The displacement at the top is 0.3*t and the bottom is fixed as shown in the attached figures. But weirdly the maximum von Mises stress is at the top not at the bottom as expected.
 
![NRB_load](https://user-images.githubusercontent.com/43910101/224035463-33b44c39-5664-4b2e-9d52-7196b6598dac.png)
![NRB_BC](https://user-images.githubusercontent.com/43910101/224035475-b0939028-7196-4ff6-92e5-6a0a27827a37.png)
![NRB_result](https://user-images.githubusercontent.com/43910101/224035482-9e397371-80da-41b1-8b8d-8ba3a60de319.png)
My input file is as follwing:
```
[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
[]

[Mesh]
    [base]
        type = FileMeshGenerator
        file = NRB_R2_complete.inp
    []
[]

[Modules/TensorMechanics/Master]
  [./all]
    strain = FINITE
    add_variables = true
    generate_output = 'max_principal_stress triaxiality_stress vonmises_stress plastic_strain_xx plastic_strain_yy plastic_strain_zz'
  [../]
[]

[Functions]
  [./hf]
    type = PiecewiseLinear
    data_file = stress_data.csv
    format = columns
  [../]
[]

[BCs]
  [./y_pull_function]
    type = FunctionDirichletBC
    variable = disp_y
    boundary = top
    function = 0.3*t
  [../]
  [./y_bot]
    type = DirichletBC
    variable = disp_y
    boundary = bottom
    value = 0.0
  [../]
[]

[Materials]
  [./elasticity_tensor]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 2e5#MPa
    poissons_ratio = 0.29
  [../]
  [./isotropic_plasticity]
    type = IsotropicPlasticityStressUpdate
    yield_stress = 431.46 #MPa
    hardening_function = hf
  [../]
  [./radial_return_stress]
    type = ComputeMultipleInelasticStress
    tangent_operator = elastic
    inelastic_models = 'isotropic_plasticity'
  [../]
  [lode_angle]
    type = ComputeLodeAngle
    outputs = exodus
  []
  [critical_crack_strain]
    type = ComputeCriticalCrackStrain
    outputs = exodus
  []
  [peeq]
    type = ComputePEEQ
    outputs = exodus
  []
  [d]
    type = ComputeDamage
    alpha = 1.0
    beta = 1.0
    critical_fracture_stress = 1204
    outputs = exodus
  []
[]

[Executioner]
  type = Transient
  solve_type = 'NEWTON'

  petsc_options = '-snes_ksp_ew'
  petsc_options_iname = '-ksp_gmres_restart'
  petsc_options_value = '101'

  line_search = 'none'

  l_max_its = 50
  nl_max_its = 50
  nl_rel_tol = 1e-8
  nl_abs_tol = 1e-6
  l_tol = 1e-9

  start_time = 0.0
  end_time = 0.5
  dt = 0.005
[]

[Outputs]
  exodus = true
  console = true
  print_perf_log = true
[]
```

**Comment by GiudGiud** [comment]
nl_abs_tol = 1e-6
this is pretty coarse. Let's tighten it and/or add automatic_scaling  to make sure we're looking at converged results

Guillaume

**Reply by RWTHLHK** [reply]
Ok, thanks. I will try it.

**Reply by RWTHLHK** [reply]
I change  nl_rel_tol = 1e-10, nl_abs_tol = 1e-12. The maximum von mises stress is still at the top...

**Reply by GiudGiud** [reply]
2e5#MPa how did you change the units here?

I think this should be in Pa

**Reply by RWTHLHK** [reply]
I didn't pay attention to the units yet. I want to see how can I incorporate the hardening law into the simulation. So I used the same data as that in an Abaqus simulation, the young modulus is 200GPa=2e5MPa.

**Reply by GiudGiud** [reply]
=2e11 Pa.
Super low young modulus in the simulation would impact the results

**Reply by RWTHLHK** [reply]
Hello, I changed the young's modulus to 2e11, the result was still the same. Then I used a simple cylinder model to test with same BCs and material properties, and simply used ComputeLinearElasticStress, the maximum von mises stress is still at the top.

**Reply by GiudGiud** [reply]
@recuero you may want to check this thread too for context

**Reply by recuero** [reply, marked as answer]
> But weirdly the maximum von Mises stress is at the top not at the bottom as expected.

If you are modeling cracking, the areas with large stresses will suddenly have zero stress. In that case, your results wouldn't look too bad depending on _what_ you are modeling. 

When you say you simply used `ComputeLinearElasticStress`: Is that all you used? Or you also accounted for damage?

**Reply by hugary1995** [reply]
I agree with @recuero this is due to your damage model.

**Reply by RWTHLHK** [reply]
> > But weirdly the maximum von Mises stress is at the top not at the bottom as expected.
> 
> If you are modeling cracking, the areas with large stresses will suddenly have zero stress. In that case, your results wouldn't look too bad depending on _what_ you are modeling.
> 
> When you say you simply used `ComputeLinearElasticStress`: Is that all you used? Or you also accounted for damage?

Yes this is all what I used. To test I disabled damage computation during the simulation, but the maximum stress is still at the top.

**Reply by hugary1995** [reply]
If so can you post the input file without your custom materials?

**Reply by RWTHLHK** [reply]
The following file is my input file without my custom materials:
`[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
[]

[Mesh]
    [base]
        type = FileMeshGenerator
        file = test.inp
    []
[]

[Modules/TensorMechanics/Master]
  [./all]
    strain = SMALL
    add_variables = true
    generate_output = 'max_principal_stress triaxiality_stress vonmises_stress'
  [../]
[]

[Functions]
  [./hf]
    type = PiecewiseLinear
    data_file = stress_data.csv
    format = columns
  [../]
[]

[BCs]
  [./y_pull_function]
    type = FunctionDirichletBC
    variable = disp_y
    boundary = top
    function = 0.3*t
  [../]
  [./y_bot]
    type = DirichletBC
    variable = disp_y
    boundary = bottom
    value = 0.0
  [../]
[]

[Materials]
  [./elasticity_tensor]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 2e11#Pa
    poissons_ratio = 0.29
  [../]
  [stress]
    type = ComputeLinearElasticStress
  []
  [density]
    type = GenericConstantMaterial
    prop_names = density
    prop_values = 7850
  []
[]

[Executioner]
  type = Transient
  solve_type = 'NEWTON'

  petsc_options = '-snes_ksp_ew'
  petsc_options_iname = '-ksp_gmres_restart'
  petsc_options_value = '101'

  line_search = 'none'

  l_max_its = 50
  nl_max_its = 50
  nl_rel_tol = 1e-10
  nl_abs_tol = 1e-12
  l_tol = 1e-9

  start_time = 0.0
  end_time = 0.5
  dt = 0.005
[]

[Outputs]
  exodus = true
  console = true
  print_perf_log = true
[]
`

**Reply by hugary1995** [reply]
Thanks. Is your stress_data.csv available? If not I can try replacing it with a simpler function.

**Reply by hugary1995** [reply]
Also your mesh please.

**Reply by RWTHLHK** [reply]
> Thanks. Is your stress_data.csv available? If not I can try replacing it with a simpler function.

Ok, thanks a lot for your help!
[problems.zip](https://github.com/idaholab/moose/files/10957979/problems.zip)

**Reply by RWTHLHK** [reply]
> Thanks. Is your stress_data.csv available? If not I can try replacing it with a simpler function.

And NRB_R2_complete.inp is the following one:
[NRB_R2_complete.inp.zip](https://github.com/idaholab/moose/files/10957988/NRB_R2_complete.inp.zip)

**Reply by hugary1995** [reply]
There are a few issues with your problem:
1. Your mesh contains both hex and tet elements. For whatever reason the tet elements are deleted by the mesh generator. Please use a hex only mesh.
2. There is volumetric locking. Set `volumetric_locking_correction = true` or use higher order elements.
3. There are rigid body modes -- you only fixed the movement in y direction.
4. I don't know how you got the solve to converge with rigid body modes. But you should consider using a direct solver if possible, if not try hypre boomeramg.
5. Use the new system which gives you exact Jacobian.
6. Your yield stress is defined in unit of MPa (my guess based on the magnitudes...) while your young's modulus is defined using Pa.
7. Use automatic scaling.
8. Use preset = false for Dirichlet BCs in plasticity simulations.

If you can fix your mesh, the following input file should be a good starting point
```python
[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
  large_kinematics = true
  volumetric_locking_correction = true
[]

[Mesh]
  [base]
    type = FileMeshGenerator
    file = NRB_R2_complete.inp
  []
[]

[Modules/TensorMechanics/Master]
  [all]
    strain = FINITE
    add_variables = true
    new_system = true
    formulation = TOTAL
  []
[]

[BCs]
  [y_pull_function]
    type = FunctionDirichletBC
    variable = disp_y
    boundary = top
    function = 0.3*t
    preset = false
  []
  [x_top]
    type = DirichletBC
    variable = disp_x
    boundary = bottom
    value = 0.0
  []
  [z_top]
    type = DirichletBC
    variable = disp_z
    boundary = bottom
    value = 0.0
  []
  [y_bot]
    type = DirichletBC
    variable = disp_y
    boundary = bottom
    value = 0.0
  []
[]

[Materials]
  [elasticity_tensor]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 2e5 #MPa
    poissons_ratio = 0.29
  []
  [flow_stress]
    type = DerivativeParsedMaterial
    property_name = flow_stress
    expression = '431+11740*effective_plastic_strain'
    material_property_names = 'effective_plastic_strain'
    additional_derivative_symbols = 'effective_plastic_strain'
    derivative_order = 2
    compute = false
  []
  [compute_stress]
    type = ComputeSimoHughesJ2PlasticityStress
    flow_stress_material = flow_stress
    output_properties = 'effective_plastic_strain'
    outputs = 'exodus'
  []
  [vonmises]
    type = RankTwoInvariant
    rank_two_tensor = cauchy_stress
    property_name = vonmises
    invariant = VonMisesStress
    outputs = 'exodus'
  []
[]

[Executioner]
  type = Transient
  solve_type = NEWTON
  petsc_options_iname = '-pc_type'
  petsc_options_value = 'lu'
  automatic_scaling = true
  line_search = 'none'
  l_max_its = 50
  nl_max_its = 50
  nl_rel_tol = 1e-6
  nl_abs_tol = 1e-8
  l_tol = 1e-9
  start_time = 0.0
  end_time = 0.5
  dt = 0.005
[]

[Outputs]
  exodus = true
  print_linear_residuals = false
[]
```

**Reply by RWTHLHK** [reply]
Thanks a lot for your reply! I regenerated the mesh using full hex elements and used your input file to run the simulation, but the max stress is still at the top...By the way, how can I check the type of mesh elements and whether they are deleted or not?

**Reply by RWTHLHK** [reply]
> There are a few issues with your problem:
> 
>     1. Your mesh contains both hex and tet elements. For whatever reason the tet elements are deleted by the mesh generator. Please use a hex only mesh.
> 
>     2. There is volumetric locking. Set `volumetric_locking_correction = true` or use higher order elements.
> 
>     3. There are rigid body modes -- you only fixed the movement in y direction.
> 
>     4. I don't know how you got the solve to converge with rigid body modes. But you should consider using a direct solver if possible, if not try hypre boomeramg.
> 
>     5. Use the new system which gives you exact Jacobian.
> 
>     6. Your yield stress is defined in unit of MPa (my guess based on the magnitudes...) while your young's modulus is defined using Pa.
> 
>     7. Use automatic scaling.
> 
>     8. Use preset = false for Dirichlet BCs in plasticity simulations.
> 
> 
> If you can fix your mesh, the following input file should be a good starting point
> 
> ```python
> [GlobalParams]
>   displacements = 'disp_x disp_y disp_z'
>   large_kinematics = true
>   volumetric_locking_correction = true
> []
> 
> [Mesh]
>   [base]
>     type = FileMeshGenerator
>     file = NRB_R2_complete.inp
>   []
> []
> 
> [Modules/TensorMechanics/Master]
>   [all]
>     strain = FINITE
>     add_variables = true
>     new_system = true
>     formulation = TOTAL
>   []
> []
> 
> [BCs]
>   [y_pull_function]
>     type = FunctionDirichletBC
>     variable = disp_y
>     boundary = top
>     function = 0.3*t
>     preset = false
>   []
>   [x_top]
>     type = DirichletBC
>     variable = disp_x
>     boundary = bottom
>     value = 0.0
>   []
>   [z_top]
>     type = DirichletBC
>     variable = disp_z
>     boundary = bottom
>     value = 0.0
>   []
>   [y_bot]
>     type = DirichletBC
>     variable = disp_y
>     boundary = bottom
>     value = 0.0
>   []
> []
> 
> [Materials]
>   [elasticity_tensor]
>     type = ComputeIsotropicElasticityTensor
>     youngs_modulus = 2e5 #MPa
>     poissons_ratio = 0.29
>   []
>   [flow_stress]
>     type = DerivativeParsedMaterial
>     property_name = flow_stress
>     expression = '431+11740*effective_plastic_strain'
>     material_property_names = 'effective_plastic_strain'
>     additional_derivative_symbols = 'effective_plastic_strain'
>     derivative_order = 2
>     compute = false
>   []
>   [compute_stress]
>     type = ComputeSimoHughesJ2PlasticityStress
>     flow_stress_material = flow_stress
>     output_properties = 'effective_plastic_strain'
>     outputs = 'exodus'
>   []
>   [vonmises]
>     type = RankTwoInvariant
>     rank_two_tensor = cauchy_stress
>     property_name = vonmises
>     invariant = VonMisesStress
>     outputs = 'exodus'
>   []
> []
> 
> [Executioner]
>   type = Transient
>   solve_type = NEWTON
>   petsc_options_iname = '-pc_type'
>   petsc_options_value = 'lu'
>   automatic_scaling = true
>   line_search = 'none'
>   l_max_its = 50
>   nl_max_its = 50
>   nl_rel_tol = 1e-6
>   nl_abs_tol = 1e-8
>   l_tol = 1e-9
>   start_time = 0.0
>   end_time = 0.5
>   dt = 0.005
> []
> 
> [Outputs]
>   exodus = true
>   print_linear_residuals = false
> []
> ```

And this is a one-half model, I need to apply symmetry boundary conditions at the bottom, I am not sure if I need to set the displacements in x-and z-direction of the bottom to 0.
